File: app-android/app/src/main/java/de/tutao/tutanota/credentials/DataKeyGeneratorBeforeAPI30.java
package de.tutao.tutanota.credentials;

import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;

import java.security.InvalidAlgorithmParameterException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;

import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;

public final class DataKeyGeneratorBeforeAPI30 implements DataKeyGenerator {
	@Override
	public SecretKey generateDataKey(String alias, CredentialEncryptionMode credentialEncryptionMode) {
		try {
			KeyGenerator keyGenerator = KeyGenerator.getInstance("AES", "AndroidKeyStore");
			KeyGenParameterSpec.Builder builder = new KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
					.setBlockModes(KeyProperties.BLOCK_MODE_CBC)
					.setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
					.setRandomizedEncryptionRequired(true)
					.setUserAuthenticationRequired(credentialEncryptionMode.requiresAuthentication);
			switch (credentialEncryptionMode) {
				// This looks odd, right?
				// see https://stackoverflow.com/questions/66067041/how-to-replace-deprecated-keygenparameterspec-builder-setuserauthenticationvalid
				case ENCRYPTION_MODE_BIOMETRICS:
					builder.setUserAuthenticationValidityDurationSeconds(-1);
					break;
				case ENCRYPTION_MODE_SYSTEM_PASSWORD:
					builder.setUserAuthenticationValidityDurationSeconds(10);
					break;
				case ENCRYPTION_MODE_DEVICE_LOCK:
					break;
			}
			keyGenerator.init(builder.build());
			return keyGenerator.generateKey();
		} catch (InvalidAlgorithmParameterException | NoSuchAlgorithmException | NoSuchProviderException e) {
			throw new RuntimeException(e);
		}
	}
}

File: app-android/app/src/main/java/de/tutao/tutanota/credentials/DataKeyGeneratorBeforeAPI30.kt
package de.tutao.tutanota.credentials

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.InvalidAlgorithmParameterException
import java.security.NoSuchAlgorithmException
import java.security.NoSuchProviderException
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

class DataKeyGeneratorBeforeAPI30 : DataKeyGenerator {
	override fun generateDataKey(alias: String, credentialEncryptionMode: CredentialEncryptionMode): SecretKey {
		return try {
			val keyGenerator = KeyGenerator.getInstance("AES", "AndroidKeyStore")
			val builder = KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
					.setBlockModes(KeyProperties.BLOCK_MODE_CBC)
					.setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
					.setRandomizedEncryptionRequired(true)
					.setUserAuthenticationRequired(credentialEncryptionMode.requiresAuthentication)
			when (credentialEncryptionMode) {
				CredentialEncryptionMode.ENCRYPTION_MODE_BIOMETRICS      -> builder.setUserAuthenticationValidityDurationSeconds(-1)
				CredentialEncryptionMode.ENCRYPTION_MODE_SYSTEM_PASSWORD -> builder.setUserAuthenticationValidityDurationSeconds(10)
				CredentialEncryptionMode.ENCRYPTION_MODE_DEVICE_LOCK     -> {}
			}
			keyGenerator.init(builder.build())
			keyGenerator.generateKey()
		} catch (e: InvalidAlgorithmParameterException) {
			throw RuntimeException(e)
		} catch (e: NoSuchAlgorithmException) {
			throw RuntimeException(e)
		} catch (e: NoSuchProviderException) {
			throw RuntimeException(e)
		}
	}
}